### PR TITLE
Vision cones

### PR DESCRIPTION
**Pull request by mikomyazaki**
Players now have a 90 degree arc behind themselves where they cannot see mobs, unless that mob is grabbing them.

[FEEDBACK THREAD](https://forums.aurorastation.org/topic/14724-vision-cones-feedback)

Intentions with this change are:
- More antagonist possibilities with sneaking about, potential to add things like pick-pocketing in the future.
- Can make some alien types have full vision to make them more unique, if appropriate to their biology.
- I guess the immersion/realism angle. 

This PR is based off of #8616 which is itself based off of [https://github.com/Baystation12/Baystation12/pull/21764](https://github.com/Baystation12/Baystation12/pull/21764)

Differences from #8616 are optimizations to greatly reduce the number of calls to InCone(), main points are:
- Only cares about turfs when position checking, instead of looping over all atoms in view.
- Does some stuff to reduce the number of turfs it has to check from all turfs in oview to all turfs in approx. half oview.

ToDo:
- [x] Make things being pulled by another mob invisible too
- [x] Fix binoculars, scopes
- [x] See about using oviewers instead
- [x] Improve processing for stationary mobs, so they don't have to recheck their whole cone when someone else moves
- [x] Hide visible emotes from invisible people.
- [x] Add a pref to change the opacity of the darkness cone
- [x] Fix pAI vision cones when they fold in and out
- [x] Make HUD icons not appear in the darkness
- [x] Stop people appearing in the darkness when they move
- [x] Laying down should update your visible mobs faster.
- [x] Ghosts have vision cones
- [x] Apparently grabs don't work properly?
- [x] Show held item works if they are behind you
- [x] Can't see mobs buckled to a bed you're pulling
- [x] pAI vision cones are wonky if you get picked up, and then wonkier when you get dropped
- [x] Fix problem with people disappearing infront of you... 
- [x] Mobs are visible on the furthest title EAST when looking WEST
- [x] Mobs are visible sometimes on the edge of your vision when moving
- [x] Ghosts no longer produce images
- [x] Make it possible to change direction via clicking while Face Direction mode is on

Optional / Future PRs / Not sure about:
- See if we can hide oily footprints from hidden mobs until you turn.
- Apparently "Can't point at people that can't see you"
- Decide how vehicles like the Janicart should work
- Apparently the lying down thing isn't fixed fully? _CANNOT REPRODUCE_
- Show an outline for people talking behind you?

### CODE REVIEW

**Comment by mikomyazaki** [comment]
!review

**Comment by Snakebittenn** [comment]
I would prefer a 120 degree arc or something similar like humans in real life. The parts you can't see of your screen are too minimal.

**Comment by mikomyazaki** [comment]
> I would prefer a 120 degree arc or something similar like humans in real life. The parts you can't see of your screen are too minimal.

Having a vision cone that is not a multiple of 90 degrees will be more expensive to calculate. I'd like to see if this version works at all with many players, before I worry about the exact size.

**Comment by Doxxmedearly** [comment]
Shadow of the cone feels a little dark. Any way to lighten it?

**Comment by mikomyazaki** [comment]
!review

**Comment by mikomyazaki** [comment]
Main bugs fixed & feedback (so far) addressed from today's test-merging.

File: code/modules/mob/living/carbon/carbon.dm [hunk 1]
```diff
@@ -435,6 +435,9 @@
 /mob/living/carbon/proc/need_breathe()
 	return
 
+/*/mob/living/carbon/human/is_invisible_to(var/mob/viewer)
```
Comment (NonQueueingMatt) [inline code/modules/mob/living/carbon/carbon.dm:438]: commented?

Comment (mikomyazaki) [reply code/modules/mob/living/carbon/carbon.dm]: Purged.

**Comment by mikomyazaki** [comment]
All reported bugs addressed up to the time of writing.

**Comment by mikomyazaki** [comment]
!review

**Comment by Chen-Yakumo** [comment]
pAIs are affected by the vision cone when in mobile form, and not when in card form. Any mobs obscured by the cone when going from mobile to card seem to stay invisible, however. Further, when being worn as a hat as a cat pAI, the vision cone was still visible on screen but ceased to effect mobs or turn with the player I was being worn by, effectively giving me old normal vision but with a weird shadow on one side.

**Comment by StationCrab** [comment]
Hello. There seems to be an issue with grabbing borgs. It may also exist with other mobs, but I experienced it as a borg.

When someone grabs you and drags you, if you look away from that person you seem to go totally blind, and everyone vanishes including yourself.

Here is a gif demonstrating this: https://i.gyazo.com/d894c4812f4c8849fb7b8735b7be3c7c.gif

**Pull request merged**